iroh-net/src/hp/portmapper.rs

```diff
@@ -14,41 +21,151 @@ pub struct ProbeResult {
 
 /// A port mapping client.
 #[derive(Default, Debug, Clone)]
-pub struct Client {}
+pub struct Client {
+    local_port: u16,
+
+    last_upnp_gateway_addr: Option<(SocketAddrV4, Instant)>,
+    current_mapping: Option<upnp::Mapping>,
+}
 
 impl Client {
     pub fn new() -> Self {
         Self::default()
     }
 
-    pub async fn probe(&self) -> Result<ProbeResult, Error> {
-        bail!("not implemented yet")
+    /// Releases the current mapping and clears it from the cache.
+    async fn invalidate_mapping(&mut self) -> Result<(), Error> {
+        if let Some(old_mapping) = self.current_mapping.take() {
+            old_mapping.release().await?;
+        }
+        Ok(())
+    }
+
+    /// UPnP: searchs for an upnp internet gateway device (a router).
+    pub async fn probe(&mut self) -> Result<ProbeResult, Error> {
+        info!("running upnp probe");
+        let upnp = self.upnp_available_from_cache() || self.probe_upnp_available().await;
+
+        info!("upnp probe result {upnp}");
+        Ok(ProbeResult {
+            pcp: false,
+            pmp: false,
+            upnp,
+        })
+    }
+
+    /// Checks if the last seen upnp gateway should still be trusted as available.
+    ///
+    /// Returns false otherwise, or if there is no cached gateway.
+    fn upnp_available_from_cache(&self) -> bool {
+        self.last_upnp_gateway_addr
+            .as_ref()
+            .map(|(_gateway_addr, last_probed)| {
+                *last_probed + AVAILABILITY_TRUST_DURATION > Instant::now()
+            })
+            .unwrap_or_default()
+    }
+
+    /// Probes for UPnP gateways.
+    async fn probe_upnp_available(&mut self) -> bool {
+        match upnp::probe_available().await {
+            Ok(gateway_addr) => {
+                let old_gateway = self
+                    .last_upnp_gateway_addr
+                    .replace((gateway_addr, Instant::now()));
+                if let Some((old_gateway_addr, _last_seen)) = old_gateway {
+                    if old_gateway_addr != gateway_addr {
+                        tracing::debug!(
+                            "upnp gateway changed from {old_gateway_addr} to {gateway_addr}"
+                        );
+                        // TODO(@divagant-martian): tailscale does not invalidate the mappings here. Why?
+                    }
+                }
+                true
+            }
+            Err(e) => {
+                info!("upnp probe failed {e}");
+                // invalidate last seen gateway and time
+                self.last_upnp_gateway_addr = None;
+                false
+            }
+        }
     }
 
     /// Updates the local port number to which we want to port map UDP traffic.
-    pub async fn set_local_port(&self, _local_port: u16) {
-        // TODO:
+    /// Invalidates the current mapping if any.
+    pub async fn set_local_port(&mut self, local_port: u16) {
+        // TODO(@divagant-martian): this should never be 0
```

Comment:
maybe use `NonZeroU16`?

Reply:
done in https://github.com/n0-computer/iroh/pull/1117/commits/4d2b14bdebd6390ca0974f01c0772d6eb26b2200